### PR TITLE
Introduce Tuttifrutti.Cache.Storage

### PR DESCRIPTION
**Pull request by zudov**
Tuttifrutti.Cache.Storage provides an interface intended for storing
cached values with a basic set of operations to CRUD and expire the
items.

The provided implementation (Tuttifrutti.Cache.Storage.InMemory)
maintains the cache in an in-memory HashPSQ. A postgres backend is
coming soon.

All of that is based on cache implementations in Faro and Persona
and should cover the needs of both.

### CODE REVIEW

**Comment by zudov** [comment]
This is review-worthy now. Currently I haven't changed `Tuttifrutti.Cache` much, mostly moved all the code to `Tuttifrutti.Cache.Storage` and left the type signatures and names in `Tuttifrutti.Cache` fairly the same. This way the diff is more reviewable.

The next step is to change the interface of `Tuttifrutti.Cache` and cover it with tests. Then I'll drop the second part of the Cache `Tuttifrutti.Cache.Nursery` which will reference cache keys before their values are available and allow to block on them until they are received. `Tuttifrutti.Cache` will then use it and its tests would test the correctness of both.

File: src/Tuttifrutti/Cache/Storage.hs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+module Tuttifrutti.Cache.Storage where
+
+import           Tuttifrutti.Prelude
+
+import           Data.Range.Range    (Range (..))
+
+-- | Storage implementation. This defines primitive operations, from which
+--   all others are derived.
+data Handle k p v m = Handle
```
Comment (f-f) [inline src/Tuttifrutti/Cache/Storage.hs]: Let's spell out `k`, `p` and `v` at least here

File: src/Tuttifrutti/Cache/Storage.hs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  { -- ^ Insert, update, or delete element under a given key.
```
Comment (f-f) [inline src/Tuttifrutti/Cache/Storage.hs]: Probably worth detailing more what this function passed as first param does

File: src/Tuttifrutti/Cache/Storage.hs (hunk 1 above)
Comment (zudov) [reply src/Tuttifrutti/Cache/Storage.hs]: I ended up not changing the parameter names, but described them in the docstring. This way all the code base would be using the same letters and their meaning is documented here.

**Pull request merged**